<?xml version="1.0" ?>
<!--
  Gazebo lookup wheel slip demo

  The world consists of two vehicles, blue and green, on a flat heightmap.
  The blue vehicle has a lookup wheel slip plugin while the green only has
  a normal wheel slip plugin with fixed slip compliance values.

  The slip map used in this demo contains several regions with increased
  longitudinal slip compliance values. The same slip map is overlaid on top
  of the heightmap to help visualize where the slippage regions are. When the
  blue vehicle drives over these green regions, its wheels' longitudinal slip
  compliance values should be dynamically updated (increased).

  Note: wheel slippage takes effect when a vehicle accelerates, e.g.
  climbing uphill. So the gravity is set to a small -x direction to simulate
  an inclined surface.

  Run with:
    gz sim -v 4 lookup_wheel_slip.sdf

  Try sending commands:
    gz topic -t "/model/vehicle_blue/cmd_vel" -m gz.msgs.Twist -p "linear: {x: 1.0}"

    gz topic -t "/model/vehicle_green/cmd_vel" -m gz.msgs.Twist -p "linear: {x: 1.0}"

  See the blue vehicle does not drive in a straight line because it's slipping
  more over the colored squares.
-->

<sdf version="1.6">
  <world name="lookup_wheel_slip">
    <!-- Set gravity to a small -x direction to simulate climbing up a hill -->
    <gravity>-1 0 -9.8</gravity>
    <physics name="1ms" type="ignored">
      <dart>
        <!-- Heightmaps behave better with the bullet collision detector -->
        <collision_detector>bullet</collision_detector>
      </dart>
    </physics>

    <plugin
      filename="gz-sim-physics-system"
      name="gz::sim::systems::Physics">
    </plugin>
    <plugin
      filename="gz-sim-user-commands-system"
      name="gz::sim::systems::UserCommands">
    </plugin>
    <plugin
      filename="gz-sim-scene-broadcaster-system"
      name="gz::sim::systems::SceneBroadcaster">
    </plugin>

    <model name="ground_plane">
      <static>true</static>
      <link name="link">
        <collision name="collision">
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>100 100</size>
            </plane>
          </geometry>
        </collision>
        <visual name="visual">
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>100 100</size>
            </plane>
          </geometry>
          <material>
            <ambient>0.8 0.8 0.8 1</ambient>
            <diffuse>0.8 0.8 0.8 1</diffuse>
            <specular>0.8 0.8 0.8 1</specular>
          </material>
        </visual>
      </link>
    </model>
    <light type="directional" name="sun">
      <cast_shadows>true</cast_shadows>
      <pose>0 0 10 0 0 0</pose>
      <diffuse>1 1 1 1</diffuse>
      <specular>0.5 0.5 0.5 1</specular>
      <attenuation>
        <range>1000</range>
        <constant>0.9</constant>
        <linear>0.01</linear>
        <quadratic>0.001</quadratic>
      </attenuation>
      <direction>-0.5 0.1 -0.9</direction>
    </light>

    <model name='vehicle_green'>
      <pose>-5.3105 -2 0.325 0 -0 0</pose>

      <link name='chassis'>
        <pose>-0.151427 -0 0.175 0 -0 0</pose>
        <inertial>
          <mass>1.14395</mass>
          <inertia>
            <ixx>0.126164</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.416519</iyy>
            <iyz>0</iyz>
            <izz>0.481014</izz>
          </inertia>
        </inertial>
        <visual name='visual'>
          <geometry>
            <box>
              <size>2.01142 1 0.568726</size>
            </box>
          </geometry>
          <material>
            <ambient>0.5 1.0 0.5 1</ambient>
            <diffuse>0.5 1.0 0.5 1</diffuse>
            <specular>0.0 1.0 0.0 1</specular>
          </material>
        </visual>
        <collision name='collision'>
          <geometry>
            <box>
              <size>2.01142 1 0.568726</size>
            </box>
          </geometry>
        </collision>
      </link>

      <link name='left_wheel'>
        <pose>0.554283 0.625029 -0.025 -1.5707 0 0</pose>
        <inertial>
          <mass>2</mass>
          <inertia>
            <ixx>0.145833</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.145833</iyy>
            <iyz>0</iyz>
            <izz>0.125</izz>
          </inertia>
        </inertial>
        <visual name='visual'>
          <geometry>
            <sphere>
              <radius>0.3</radius>
            </sphere>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name='collision'>
          <geometry>
            <sphere>
              <radius>0.3</radius>
            </sphere>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>1</mu>
                <mu2>1</mu2>
                <slip1>0.035</slip1>
                <slip2>0</slip2>
                <fdir1>0 0 1</fdir1>
              </ode>
              <bullet>
                <friction>1</friction>
                <friction2>1</friction2>
                <rolling_friction>0.1</rolling_friction>
              </bullet>
            </friction>
          </surface>
        </collision>
      </link>

      <link name='right_wheel'>
        <pose>0.554282 -0.625029 -0.025 -1.5707 0 0</pose>
        <inertial>
          <mass>2</mass>
          <inertia>
            <ixx>0.145833</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.145833</iyy>
            <iyz>0</iyz>
            <izz>0.125</izz>
          </inertia>
        </inertial>
        <visual name='visual'>
          <geometry>
            <sphere>
              <radius>0.3</radius>
            </sphere>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name='collision'>
          <geometry>
            <sphere>
              <radius>0.3</radius>
            </sphere>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>1</mu>
                <mu2>1</mu2>
                <slip1>0.035</slip1>
                <slip2>0</slip2>
                <fdir1>0 0 1</fdir1>
              </ode>
              <bullet>
                <friction>1</friction>
                <friction2>1</friction2>
                <rolling_friction>0.1</rolling_friction>
              </bullet>
            </friction>
          </surface>
        </collision>
      </link>

      <link name='caster'>
        <pose>-0.957138 -0 -0.125 0 -0 0</pose>
        <inertial>
          <mass>1</mass>
          <inertia>
            <ixx>0.1</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.1</iyy>
            <iyz>0</iyz>
            <izz>0.1</izz>
          </inertia>
        </inertial>
        <visual name='visual'>
          <geometry>
            <sphere>
              <radius>0.2</radius>
            </sphere>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name='collision'>
          <geometry>
            <sphere>
              <radius>0.2</radius>
            </sphere>
          </geometry>
        </collision>
      </link>

      <joint name='left_wheel_joint' type='revolute'>
        <parent>chassis</parent>
        <child>left_wheel</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-1.79769e+308</lower>
            <upper>1.79769e+308</upper>
          </limit>
        </axis>
      </joint>

      <joint name='right_wheel_joint' type='revolute'>
        <parent>chassis</parent>
        <child>right_wheel</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-1.79769e+308</lower>
            <upper>1.79769e+308</upper>
          </limit>
        </axis>
      </joint>

      <joint name='caster_wheel' type='ball'>
        <parent>chassis</parent>
        <child>caster</child>
      </joint>

      <plugin
        filename="gz-sim-diff-drive-system"
        name="gz::sim::systems::DiffDrive">
        <left_joint>left_wheel_joint</left_joint>
        <right_joint>right_wheel_joint</right_joint>
        <wheel_separation>1.25</wheel_separation>
        <wheel_radius>0.3</wheel_radius>
        <max_linear_acceleration>1</max_linear_acceleration>
        <min_linear_acceleration>-1</min_linear_acceleration>
        <max_angular_acceleration>2</max_angular_acceleration>
        <min_angular_acceleration>-2</min_angular_acceleration>
        <max_linear_velocity>0.5</max_linear_velocity>
        <min_linear_velocity>-0.5</min_linear_velocity>
        <max_angular_velocity>1</max_angular_velocity>
        <min_angular_velocity>-1</min_angular_velocity>
      </plugin>

      <plugin
        filename="gz-sim-wheel-slip-system"
        name="gz::sim::systems::WheelSlip">
        <wheel link_name="right_wheel">
          <slip_compliance_lateral>1</slip_compliance_lateral>
          <slip_compliance_longitudinal>1</slip_compliance_longitudinal>
          <wheel_normal_force>77</wheel_normal_force>
          <wheel_radius>0.15</wheel_radius>
        </wheel>
        <wheel link_name="left_wheel">
          <slip_compliance_lateral>1</slip_compliance_lateral>
          <slip_compliance_longitudinal>1</slip_compliance_longitudinal>
          <wheel_normal_force>32</wheel_normal_force>
          <wheel_radius>0.15</wheel_radius>
        </wheel>
        <wheel link_name="caster">
          <slip_compliance_lateral>1</slip_compliance_lateral>
          <slip_compliance_longitudinal>1</slip_compliance_longitudinal>
          <wheel_normal_force>32</wheel_normal_force>
          <wheel_radius>0.15</wheel_radius>
        </wheel>
      </plugin>
    </model>

    <model name='vehicle_blue'>
      <pose>-5.3313 3.9740 0.325 0 -0 0</pose>

      <link name='chassis'>
        <pose>-0.151427 -0 0.175 0 -0 0</pose>
        <inertial>
          <mass>1.14395</mass>
          <inertia>
            <ixx>0.126164</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.416519</iyy>
            <iyz>0</iyz>
            <izz>0.481014</izz>
          </inertia>
        </inertial>
        <visual name='visual'>
          <geometry>
            <box>
              <size>2.01142 1 0.568726</size>
            </box>
          </geometry>
          <material>
            <ambient>0.5 0.5 1.0 1</ambient>
            <diffuse>0.5 0.5 1.0 1</diffuse>
            <specular>0.0 0.0 1.0 1</specular>
          </material>
        </visual>
        <collision name='collision'>
          <geometry>
            <box>
              <size>2.01142 1 0.568726</size>
            </box>
          </geometry>
        </collision>
      </link>

      <link name='left_wheel'>
        <pose>0.554283 0.625029 -0.025 -1.5707 0 0</pose>
        <inertial>
          <mass>2</mass>
          <inertia>
            <ixx>0.145833</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.145833</iyy>
            <iyz>0</iyz>
            <izz>0.125</izz>
          </inertia>
        </inertial>
        <visual name='visual'>
          <geometry>
            <sphere>
              <radius>0.3</radius>
            </sphere>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name='collision'>
          <geometry>
            <sphere>
              <radius>0.3</radius>
            </sphere>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>1</mu>
                <mu2>1</mu2>
                <slip1>0.035</slip1>
                <slip2>0</slip2>
                <fdir1>0 0 1</fdir1>
              </ode>
              <bullet>
                <friction>1</friction>
                <friction2>1</friction2>
                <rolling_friction>0.1</rolling_friction>
              </bullet>
            </friction>
          </surface>
        </collision>
      </link>

      <link name='right_wheel'>
        <pose>0.554282 -0.625029 -0.025 -1.5707 0 0</pose>
        <inertial>
          <mass>2</mass>
          <inertia>
            <ixx>0.145833</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.145833</iyy>
            <iyz>0</iyz>
            <izz>0.125</izz>
          </inertia>
        </inertial>
        <visual name='visual'>
          <geometry>
            <sphere>
              <radius>0.3</radius>
            </sphere>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name='collision'>
          <geometry>
            <sphere>
              <radius>0.3</radius>
            </sphere>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>1</mu>
                <mu2>1</mu2>
                <slip1>0.035</slip1>
                <slip2>0</slip2>
                <fdir1>0 0 1</fdir1>
              </ode>
              <bullet>
                <friction>1</friction>
                <friction2>1</friction2>
                <rolling_friction>0.1</rolling_friction>
              </bullet>
            </friction>
          </surface>
        </collision>
      </link>

      <link name='caster'>
        <pose>-0.957138 -0 -0.125 0 -0 0</pose>
        <inertial>
          <mass>1</mass>
          <inertia>
            <ixx>0.1</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.1</iyy>
            <iyz>0</iyz>
            <izz>0.1</izz>
          </inertia>
        </inertial>
        <visual name='visual'>
          <geometry>
            <sphere>
              <radius>0.2</radius>
            </sphere>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name='collision'>
          <geometry>
            <sphere>
              <radius>0.2</radius>
            </sphere>
          </geometry>
        </collision>
      </link>

      <joint name='left_wheel_joint' type='revolute'>
        <parent>chassis</parent>
        <child>left_wheel</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-1.79769e+308</lower>
            <upper>1.79769e+308</upper>
          </limit>
        </axis>
      </joint>

      <joint name='right_wheel_joint' type='revolute'>
        <parent>chassis</parent>
        <child>right_wheel</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-1.79769e+308</lower>
            <upper>1.79769e+308</upper>
          </limit>
        </axis>
      </joint>

      <joint name='caster_wheel' type='ball'>
        <parent>chassis</parent>
        <child>caster</child>
      </joint>

      <plugin
        filename="gz-sim-diff-drive-system"
        name="gz::sim::systems::DiffDrive">
        <left_joint>left_wheel_joint</left_joint>
        <right_joint>right_wheel_joint</right_joint>
        <wheel_separation>1.25</wheel_separation>
        <wheel_radius>0.3</wheel_radius>
        <odom_publish_frequency>1</odom_publish_frequency>
        <max_linear_acceleration>1</max_linear_acceleration>
        <min_linear_acceleration>-1</min_linear_acceleration>
        <max_angular_acceleration>2</max_angular_acceleration>
        <min_angular_acceleration>-2</min_angular_acceleration>
        <max_linear_velocity>0.5</max_linear_velocity>
        <min_linear_velocity>-0.5</min_linear_velocity>
        <max_angular_velocity>1</max_angular_velocity>
        <min_angular_velocity>-1</min_angular_velocity>
      </plugin>

      <plugin
        filename="gz-sim-wheel-slip-system"
        name="gz::sim::systems::WheelSlip">
        <wheel link_name="right_wheel">
          <slip_compliance_lateral>0.0</slip_compliance_lateral>
          <slip_compliance_longitudinal>0.0</slip_compliance_longitudinal>
          <wheel_normal_force>1</wheel_normal_force>
          <wheel_radius>0.3</wheel_radius>
        </wheel>
        <wheel link_name="left_wheel">
          <slip_compliance_lateral>0.0</slip_compliance_lateral>
          <slip_compliance_longitudinal>0.0</slip_compliance_longitudinal>
          <wheel_normal_force>1</wheel_normal_force>
          <wheel_radius>0.3</wheel_radius>
        </wheel>
      </plugin>
      <plugin
        filename="gz-sim-lookup-wheel-slip-system"
        name="gz::sim::systems::LookupWheelSlip">
        <slip_map>lookup_slip_map.png</slip_map>
        <size_x>129</size_x>
        <size_y>129</size_y>
        <slip_compliance_lateral_delta>0.05</slip_compliance_lateral_delta>
        <slip_compliance_longitudinal_delta>0.001</slip_compliance_longitudinal_delta>
        <friction_delta>0.5</friction_delta>
        <wheel_link_name>left_wheel</wheel_link_name>
        <wheel_link_name>right_wheel</wheel_link_name>
      </plugin>
    </model>

    <model name="heightmap">
        <static>true</static>
        <link name="link">
            <collision name="collision">
                <geometry>
                    <heightmap>
                        <uri>lookup_slip_heightmap.png</uri>
                        <size>129 129 0.001</size>
                        <pos>0 0 0</pos>
                    </heightmap>
                </geometry>
            </collision>
            <visual name="visual">
                <geometry>
                    <heightmap>
                        <use_terrain_paging>false</use_terrain_paging>
                        <texture>
                            <diffuse>lookup_slip_map.png</diffuse>
                            <normal>lookup_slip_heightmap_normal.png</normal>
                            <size>129</size>
                        </texture>
                        <blend>
                            <min_height>2</min_height>
                            <fade_dist>5</fade_dist>
                        </blend>
                        <blend>
                            <min_height>4</min_height>
                            <fade_dist>5</fade_dist>
                        </blend>
                        <uri>lookup_slip_map.png</uri>
                        <size>129 129 0.001</size>
                        <pos>0 0 0</pos>
                    </heightmap>
                </geometry>
            </visual>
        </link>
    </model>

  </world>
</sdf>
